File: src/core/search.test.ts
import { afterEach, describe, expect, test } from 'bun:test';
import {
  CURRENT_EMBEDDING_VERSION,
  CURRENT_EXTRACTION_VERSION,
  initDatabase,
  insertMemoryRecord,
  insertMemoryRecordVector,
} from './db.js';
import { __setModelForTests } from './embeddings.js';
import { search } from './search.js';

let db: ReturnType<typeof initDatabase> | null = null;

function memory(overrides: Partial<Parameters<typeof insertMemoryRecord>[1]> = {}): Parameters<typeof insertMemoryRecord>[1] {
  const archivePath = overrides.archivePath ?? '/archive/a.jsonl';
  const text = overrides.text ?? 'The transcript archive is the source of truth.';
  return {
    kind: 'fact',
    text,
    sourceKind: 'claude-projects',
    archivePath,
    lineStart: 4,
    lineEnd: 8,
    observedAt: Date.UTC(2026, 5, 1),
    project: 'memmem',
    dedupeKey: `${archivePath}:${overrides.lineStart ?? 4}:${text}`,
    extractionVersion: CURRENT_EXTRACTION_VERSION,
    embeddingVersion: CURRENT_EMBEDDING_VERSION,
    ...overrides,
  };
}

afterEach(() => {
  db?.close();
  db = null;
  __setModelForTests(null, null);
});

describe('memory search', () => {
  test('returns compact memory records from vector search without snippets', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const id = insertMemoryRecord(db, memory());
    insertMemoryRecordVector(db, id, Array.from({ length: 384 }, () => 0.1));

    const results = await search('source of truth', { db, limit: 5 });

    expect(results[0]).toEqual({
      id,
      kind: 'fact',
      text: 'The transcript archive is the source of truth.',
      sourceKind: 'claude-projects',
      archivePath: '/archive/a.jsonl',
      lineStart: 4,
      lineEnd: 8,
      observedAt: Date.UTC(2026, 5, 1),
      project: 'memmem',
      score: expect.any(Number),
    });
    expect(results[0]).not.toHaveProperty('snippet');
  });

  test('returns vector results and text fallback results without duplicates', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const vectorId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/a.jsonl',
      text: 'Semantic memory vector result.',
      project: 'alpha',
      observedAt: Date.UTC(2026, 4, 26),
      dedupeKey: 'vector-result',
    }));
    insertMemoryRecordVector(db, vectorId, Array.from({ length: 384 }, () => 0.1));

    insertMemoryRecord(db, memory({
      archivePath: '/archive/codex-sessions/b.jsonl',
      sourceKind: 'codex-sessions',
      text: 'Exact phrase search text result.',
      project: 'beta',
      observedAt: Date.UTC(2026, 4, 27),
      dedupeKey: 'fallback-result',
    }));

    const results = await search('exact phrase', { db, limit: 10 });

    expect(results.map(result => result.archivePath)).toContain('/archive/codex-sessions/b.jsonl');
    expect(new Set(results.map(result => result.id)).size).toBe(results.length);
  });

  test('filters by source kind and date', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/a.jsonl',
      text: 'Filter me old.',
      observedAt: Date.UTC(2026, 4, 25),
      dedupeKey: 'old-filter',
    }));
    insertMemoryRecord(db, memory({
      archivePath: '/archive/codex-sessions/b.jsonl',
      sourceKind: 'codex-sessions',
      text: 'Filter me new.',
      observedAt: Date.UTC(2026, 4, 26),
      dedupeKey: 'new-filter',
    }));

    const results = await search('filter me', { db, limit: 10, after: '2026-05-26', sourceKind: 'codex-sessions' });

    expect(results).toHaveLength(1);
    expect(results[0].sourceKind).toBe('codex-sessions');
  });

  test('filters text results by project', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const alphaId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/alpha.jsonl',
      text: 'Project text query alpha.',
      project: 'alpha',
      dedupeKey: 'alpha-text',
    }));
    insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/beta.jsonl',
      text: 'Project text query beta.',
      project: 'beta',
      dedupeKey: 'beta-text',
    }));

    const results = await search('project text query', { db, limit: 10, projects: ['alpha'] });

    expect(results.map(result => result.id)).toEqual([alphaId]);
  });

  test('filters vector results by project', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const alphaId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/alpha-vector.jsonl',
      text: 'Alpha semantic only.',
      project: 'alpha',
      dedupeKey: 'alpha-vector',
    }));
    insertMemoryRecordVector(db, alphaId, Array.from({ length: 384 }, () => 0.1));
    const betaId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/beta-vector.jsonl',
      text: 'Beta semantic only.',
      project: 'beta',
      dedupeKey: 'beta-vector',
    }));
    insertMemoryRecordVector(db, betaId, Array.from({ length: 384 }, () => 0.2));

    const results = await search('project-vector-query', { db, limit: 10, projects: ['beta'] });

    expect(results.map(result => result.id)).toEqual([betaId]);
  });

  test('normalizes query for vector search and text fallback', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    let embeddedQuery = '';
    __setModelForTests(async () => {}, async (_kind, text: string) => {
      embeddedQuery = text;
      return Array.from({ length: 384 }, () => 0.1);
    });
    let prompt = '';
    const queryNormalizerProvider = {
      complete: async (input: string) => {
        prompt = input;
        return { text: 'english normalized', usage: { input_tokens: 1, output_tokens: 1 } };
      },
    };

    const id = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/normalized.jsonl',
      text: 'English normalized text fallback.',
      project: null,
      dedupeKey: 'normalized',
    }));

    const results = await search('한국어 질의', { db, limit: 10, queryNormalizerProvider });

    expect(prompt).toContain('한국어 질의');
    expect(embeddedQuery).toBe('english normalized');
    expect(results.map(result => result.id)).toEqual([id]);
  });

  test('treats LIKE wildcards as literal characters', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const literalId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/literal.jsonl',
      text: 'Literal 100% match.',
      project: null,
      dedupeKey: 'literal-percent',
    }));
    insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/wildcard.jsonl',
      text: 'Literal 100abc match.',
      project: null,
      dedupeKey: 'wildcard-percent',
    }));

    const results = await search('100%', { db, limit: 10 });

    expect(results.map(result => result.id)).toEqual([literalId]);
  });

  test('finds filtered vector result outside unfiltered top k', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    for (let i = 0; i < 2; i++) {
      const id = insertMemoryRecord(db, memory({
        archivePath: `/archive/claude-projects/closer-${i}.jsonl`,
        text: 'Closer unrelated.',
        dedupeKey: `closer-${i}`,
      }));
      insertMemoryRecordVector(db, id, Array.from({ length: 384 }, () => 0.1));
    }

    const codexId = insertMemoryRecord(db, memory({
      archivePath: '/archive/codex-sessions/filtered.jsonl',
      sourceKind: 'codex-sessions',
      text: 'Codex semantic only match.',
      dedupeKey: 'codex-filtered-vector',
    }));
    insertMemoryRecordVector(db, codexId, Array.from({ length: 384 }, () => 0.2));

    const results = await search('vector-filter-query', { db, limit: 2, sourceKind: 'codex-sessions' });

    expect(results.map(result => result.id)).toEqual([codexId]);
  });

  test('returns active vector result when closer superseded vector fills top k', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const supersededId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/superseded-vector.jsonl',
      text: 'Superseded semantic only memory.',
      status: 'superseded',
      dedupeKey: 'superseded-vector-top-k',
    }));
    insertMemoryRecordVector(db, supersededId, Array.from({ length: 384 }, () => 0.1));

    const activeId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/active-vector.jsonl',
      text: 'Active semantic only memory.',
      dedupeKey: 'active-vector-after-superseded',
    }));
    insertMemoryRecordVector(db, activeId, Array.from({ length: 384 }, () => 0.2));

    const results = await search('vector-only-query', { db, limit: 1 });

    expect(results.map(result => result.id)).toEqual([activeId]);
  });

  test('preserves null project and observedAt without legacy title or snippet', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/nulls.jsonl',
      text: 'Null shape result.',
      lineStart: 5,
      lineEnd: 6,
      observedAt: null,
      project: null,
      dedupeKey: 'null-shape',
    }));

    const results = await search('null shape', { db, limit: 10 });

    expect(results).toHaveLength(1);
    expect(results[0].project).toBeNull();
    expect(results[0].observedAt).toBeNull();
    expect(Object.keys(results[0]).sort()).toEqual([
      'archivePath',
      'id',
      'kind',
      'lineEnd',
      'lineStart',
      'observedAt',
      'project',
      'sourceKind',
      'text',
    ]);
    expect(results[0]).not.toHaveProperty('title');
    expect(results[0]).not.toHaveProperty('snippet');
  });

  test('omits superseded records and truncates long memory text', async () => {
    process.env.TEST_DB_PATH = ':memory:';
    db = initDatabase();
    __setModelForTests(async () => {}, async (_kind, _text) => Array.from({ length: 384 }, () => 0.1));

    const longText = `long query ${'a'.repeat(500)}`;
    const activeId = insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/long.jsonl',
      text: longText,
      dedupeKey: 'long-active',
    }));
    insertMemoryRecord(db, memory({
      archivePath: '/archive/claude-projects/superseded.jsonl',
      text: 'long query superseded',
      status: 'superseded',
      dedupeKey: 'long-superseded',
    }));

    const results = await search('long query', { db, limit: 10 });

    expect(results.map(result => result.id)).toEqual([activeId]);
    expect(results[0].text).toHaveLength(400);
    expect(results[0].text.endsWith('...')).toBe(true);
  });
});

File: internal/core/search/search_test.go
package search

import (
	"context"
	"database/sql"
	"math"
	"testing"
	"time"

	"github.com/baleen37/memmem/internal/core/db"
)

// These tests insert vectors directly via the db package (WASM, CGO-free) and
// inject a mock Embedder, so they exercise the vector/text SQL, filters,
// combine/dedupe, and AND-intersection WITHOUT the embedding model. They port
// src/core/search.test.ts.

// constEmbedder returns the same vector for every query (the TS
// __setModelForTests([0.1]*384) pattern).
func constEmbedder(vec []float32) Embedder {
	return func(string) ([]float32, error) { return vec, nil }
}

func fillVec(v float32) []float32 {
	out := make([]float32, db.EmbeddingDim)
	for i := range out {
		out[i] = v
	}
	return out
}

// hotVector mirrors the TS hotVector helper: index hot = 1.0, rest 0.0.
func hotVector(hot int) []float32 {
	out := make([]float32, db.EmbeddingDim)
	out[hot] = 1.0
	return out
}

func observedAt(y int, mo time.Month, d int) *int64 {
	v := time.Date(y, mo, d, 0, 0, 0, 0, time.UTC).UnixMilli()
	return &v
}

func strPtr(s string) *string { return &s }

type memOpt func(*db.MemoryRecordInsert)

func memory(opts ...memOpt) db.MemoryRecordInsert {
	embV := int64(db.CurrentEmbeddingVersion)
	r := db.MemoryRecordInsert{
		Kind:              db.KindFact,
		Text:              "The transcript archive is the source of truth.",
		SourceKind:        "claude-projects",
		ArchivePath:       "/archive/a.jsonl",
		LineStart:         4,
		LineEnd:           8,
		ObservedAt:        observedAt(2026, 6, 1),
		Project:           strPtr("memmem"),
		ProjectName:       nil,
		Status:            db.StatusActive,
		DedupeKey:         "default",
		ExtractionVersion: db.CurrentExtractionVersion,
		EmbeddingVersion:  &embV,
	}
	for _, o := range opts {
		o(&r)
	}
	return r
}

func withText(s string) memOpt        { return func(r *db.MemoryRecordInsert) { r.Text = s } }
func withArchive(s string) memOpt     { return func(r *db.MemoryRecordInsert) { r.ArchivePath = s } }
func withDedupe(s string) memOpt      { return func(r *db.MemoryRecordInsert) { r.DedupeKey = s } }
func withSourceKind(s string) memOpt  { return func(r *db.MemoryRecordInsert) { r.SourceKind = s } }
func withProject(s *string) memOpt    { return func(r *db.MemoryRecordInsert) { r.Project = s } }
func withStatus(s db.MemoryRecordStatus) memOpt {
	return func(r *db.MemoryRecordInsert) { r.Status = s }
}
func withObserved(p *int64) memOpt { return func(r *db.MemoryRecordInsert) { r.ObservedAt = p } }
func withLines(start, end int64) memOpt {
	return func(r *db.MemoryRecordInsert) { r.LineStart, r.LineEnd = start, end }
}

func newTestDB(t *testing.T) *sql.DB {
	t.Helper()
	d, err := db.InitDatabase(":memory:")
	if err != nil {
		t.Fatalf("init db: %v", err)
	}
	t.Cleanup(func() { d.Close() })
	return d
}

func insertRecord(t *testing.T, d *sql.DB, r db.MemoryRecordInsert) int64 {
	t.Helper()
	id, err := db.InsertMemoryRecord(d, r)
	if err != nil {
		t.Fatalf("insert record: %v", err)
	}
	return id
}

func insertVector(t *testing.T, d *sql.DB, id int64, vec []float32) {
	t.Helper()
	if err := db.InsertMemoryRecordVector(d, id, vec); err != nil {
		t.Fatalf("insert vector: %v", err)
	}
}

func ids(results []MemorySearchResult) []int64 {
	out := make([]int64, len(results))
	for i, r := range results {
		out[i] = r.ID
	}
	return out
}

func TestSearchReturnsCompactVectorRecord(t *testing.T) {
	d := newTestDB(t)
	id := insertRecord(t, d, memory())
	insertVector(t, d, id, fillVec(0.1))

	opts := Options{DB: d, Limit: 5, Embedder: constEmbedder(fillVec(0.1))}
	results, err := Search(context.Background(), "source of truth", opts)
	if err != nil {
		t.Fatal(err)
	}
	if len(results) == 0 {
		t.Fatal("no results")
	}
	r := results[0]
	if r.ID != id || r.Kind != "fact" || r.SourceKind != "claude-projects" ||
		r.ArchivePath != "/archive/a.jsonl" || r.LineStart != 4 || r.LineEnd != 8 {
		t.Fatalf("unexpected fields: %+v", r)
	}
	if r.Text != "The transcript archive is the source of truth." {
		t.Fatalf("text = %q", r.Text)
	}
	if r.Project == nil || *r.Project != "memmem" {
		t.Fatalf("project = %v", r.Project)
	}
	if r.ObservedAt == nil || *r.ObservedAt != *observedAt(2026, 6, 1) {
		t.Fatalf("observedAt = %v", r.ObservedAt)
	}
	if r.Score == nil {
		t.Fatal("vector result must have a score")
	}
}

func TestSearchVectorAndTextNoDuplicates(t *testing.T) {
	d := newTestDB(t)
	vectorID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/a.jsonl"),
		withText("Semantic memory vector result."),
		withProject(strPtr("alpha")),
		withObserved(observedAt(2026, 5, 26)),
		withDedupe("vector-result"),
	))
	insertVector(t, d, vectorID, fillVec(0.1))

	insertRecord(t, d, memory(
		withArchive("/archive/codex-sessions/b.jsonl"),
		withSourceKind("codex-sessions"),
		withText("Exact phrase search text result."),
		withProject(strPtr("beta")),
		withObserved(observedAt(2026, 5, 27)),
		withDedupe("fallback-result"),
	))

	results, err := Search(context.Background(), "exact phrase",
		Options{DB: d, Limit: 10, Embedder: constEmbedder(fillVec(0.1))})
	if err != nil {
		t.Fatal(err)
	}
	var foundCodex bool
	seen := map[int64]bool{}
	for _, r := range results {
		if r.ArchivePath == "/archive/codex-sessions/b.jsonl" {
			foundCodex = true
		}
		if seen[r.ID] {
			t.Fatalf("duplicate id %d", r.ID)
		}
		seen[r.ID] = true
	}
	if !foundCodex {
		t.Fatal("expected text-fallback codex record in results")
	}
}

func TestSearchFiltersBySourceKindAndDate(t *testing.T) {
	d := newTestDB(t)
	insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/a.jsonl"),
		withText("Filter me old."),
		withObserved(observedAt(2026, 5, 25)),
		withDedupe("old-filter"),
	))
	insertRecord(t, d, memory(
		withArchive("/archive/codex-sessions/b.jsonl"),
		withSourceKind("codex-sessions"),
		withText("Filter me new."),
		withObserved(observedAt(2026, 5, 26)),
		withDedupe("new-filter"),
	))

	results, err := Search(context.Background(), "filter me", Options{
		DB: d, Limit: 10, After: "2026-05-26", SourceKind: "codex-sessions",
		Embedder: constEmbedder(fillVec(0.1)),
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	if results[0].SourceKind != "codex-sessions" {
		t.Fatalf("sourceKind = %q", results[0].SourceKind)
	}
}

func TestSearchFiltersTextByProject(t *testing.T) {
	d := newTestDB(t)
	alphaID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/alpha.jsonl"),
		withText("Project text query alpha."),
		withProject(strPtr("alpha")),
		withDedupe("alpha-text"),
	))
	insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/beta.jsonl"),
		withText("Project text query beta."),
		withProject(strPtr("beta")),
		withDedupe("beta-text"),
	))

	results, err := Search(context.Background(), "project text query", Options{
		DB: d, Limit: 10, Projects: []string{"alpha"}, Embedder: constEmbedder(fillVec(0.1)),
	})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != alphaID {
		t.Fatalf("ids = %v, want [%d]", got, alphaID)
	}
}

func TestSearchFiltersVectorByProject(t *testing.T) {
	d := newTestDB(t)
	alphaID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/alpha-vector.jsonl"),
		withText("Alpha semantic only."),
		withProject(strPtr("alpha")),
		withDedupe("alpha-vector"),
	))
	insertVector(t, d, alphaID, fillVec(0.1))
	betaID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/beta-vector.jsonl"),
		withText("Beta semantic only."),
		withProject(strPtr("beta")),
		withDedupe("beta-vector"),
	))
	insertVector(t, d, betaID, fillVec(0.2))

	results, err := Search(context.Background(), "project-vector-query", Options{
		DB: d, Limit: 10, Projects: []string{"beta"}, Embedder: constEmbedder(fillVec(0.1)),
	})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != betaID {
		t.Fatalf("ids = %v, want [%d]", got, betaID)
	}
}

// mockNormalizer records the prompt and returns a fixed normalization.
type mockNormalizer struct {
	prompt string
	out    string
	err    error
}

func (m *mockNormalizer) Complete(_ context.Context, prompt string) (string, error) {
	m.prompt = prompt
	return m.out, m.err
}

func TestSearchNormalizesQuery(t *testing.T) {
	d := newTestDB(t)
	id := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/normalized.jsonl"),
		withText("English normalized text fallback."),
		withProject(nil),
		withDedupe("normalized"),
	))

	var embeddedQuery string
	embedder := func(text string) ([]float32, error) {
		embeddedQuery = text
		return fillVec(0.1), nil
	}
	norm := &mockNormalizer{out: "english normalized"}

	results, err := Search(context.Background(), "한국어 질의", Options{
		DB: d, Limit: 10, QueryNormalizer: norm, Embedder: embedder,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !contains(norm.prompt, "한국어 질의") {
		t.Fatalf("prompt did not contain raw query: %q", norm.prompt)
	}
	if embeddedQuery != "english normalized" {
		t.Fatalf("embedded query = %q, want normalized", embeddedQuery)
	}
	if got := ids(results); len(got) != 1 || got[0] != id {
		t.Fatalf("ids = %v, want [%d]", got, id)
	}
}

func TestSearchLikeWildcardsLiteral(t *testing.T) {
	d := newTestDB(t)
	literalID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/literal.jsonl"),
		withText("Literal 100% match."),
		withProject(nil),
		withDedupe("literal-percent"),
	))
	insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/wildcard.jsonl"),
		withText("Literal 100abc match."),
		withProject(nil),
		withDedupe("wildcard-percent"),
	))

	results, err := Search(context.Background(), "100%",
		Options{DB: d, Limit: 10, Embedder: constEmbedder(fillVec(0.1))})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != literalID {
		t.Fatalf("ids = %v, want [%d] (literal only)", got, literalID)
	}
}

func TestSearchFiltersVectorOutsideTopK(t *testing.T) {
	d := newTestDB(t)
	for i := 0; i < 2; i++ {
		id := insertRecord(t, d, memory(
			withArchive("/archive/claude-projects/closer.jsonl"),
			withText("Closer unrelated."),
			withDedupe("closer-"+string(rune('a'+i))),
		))
		insertVector(t, d, id, fillVec(0.1))
	}
	codexID := insertRecord(t, d, memory(
		withArchive("/archive/codex-sessions/filtered.jsonl"),
		withSourceKind("codex-sessions"),
		withText("Codex semantic only match."),
		withDedupe("codex-filtered-vector"),
	))
	insertVector(t, d, codexID, fillVec(0.2))

	results, err := Search(context.Background(), "vector-filter-query", Options{
		DB: d, Limit: 2, SourceKind: "codex-sessions", Embedder: constEmbedder(fillVec(0.1)),
	})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != codexID {
		t.Fatalf("ids = %v, want [%d]", got, codexID)
	}
}

func TestSearchExcludesSupersededVectorInTopK(t *testing.T) {
	d := newTestDB(t)
	supersededID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/superseded-vector.jsonl"),
		withText("Superseded semantic only memory."),
		withStatus(db.StatusSuperseded),
		withDedupe("superseded-vector-top-k"),
	))
	insertVector(t, d, supersededID, fillVec(0.1))
	activeID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/active-vector.jsonl"),
		withText("Active semantic only memory."),
		withDedupe("active-vector-after-superseded"),
	))
	insertVector(t, d, activeID, fillVec(0.2))

	results, err := Search(context.Background(), "vector-only-query",
		Options{DB: d, Limit: 1, Embedder: constEmbedder(fillVec(0.1))})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != activeID {
		t.Fatalf("ids = %v, want [%d] (active, superseded excluded)", got, activeID)
	}
}

func TestSearchPreservesNullProjectAndObserved(t *testing.T) {
	d := newTestDB(t)
	insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/nulls.jsonl"),
		withText("Null shape result."),
		withLines(5, 6),
		withObserved(nil),
		withProject(nil),
		withDedupe("null-shape"),
	))

	results, err := Search(context.Background(), "null shape",
		Options{DB: d, Limit: 10, Embedder: constEmbedder(fillVec(0.1))})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 1 {
		t.Fatalf("expected 1 result, got %d", len(results))
	}
	if results[0].Project != nil {
		t.Errorf("project should be nil, got %v", *results[0].Project)
	}
	if results[0].ObservedAt != nil {
		t.Errorf("observedAt should be nil, got %v", *results[0].ObservedAt)
	}
	// Text-path result carries no score.
	if results[0].Score != nil {
		t.Errorf("text result should have no score, got %v", *results[0].Score)
	}
}

func TestSearchOmitsSupersededAndTruncatesText(t *testing.T) {
	d := newTestDB(t)
	longText := "long query " + repeat("a", 500)
	activeID := insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/long.jsonl"),
		withText(longText),
		withDedupe("long-active"),
	))
	insertRecord(t, d, memory(
		withArchive("/archive/claude-projects/superseded.jsonl"),
		withText("long query superseded"),
		withStatus(db.StatusSuperseded),
		withDedupe("long-superseded"),
	))

	results, err := Search(context.Background(), "long query",
		Options{DB: d, Limit: 10, Embedder: constEmbedder(fillVec(0.1))})
	if err != nil {
		t.Fatal(err)
	}
	if got := ids(results); len(got) != 1 || got[0] != activeID {
		t.Fatalf("ids = %v, want [%d]", got, activeID)
	}
	if r := []rune(results[0].Text); len(r) != 400 {
		t.Fatalf("truncated text rune length = %d, want 400", len(r))
	}
	if !endsWith(results[0].Text, "...") {
		t.Fatalf("expected trailing ..., got %q", results[0].Text)
	}
}

func TestScoreIsInverseDistance(t *testing.T) {
	d := newTestDB(t)
	// Record vector equal to query => distance 0 => score 1/(1+0) = 1.
	id := insertRecord(t, d, memory(withText("exact match"), withDedupe("exact")))
	insertVector(t, d, id, hotVector(0))

	results, err := Search(context.Background(), "q",
		Options{DB: d, Limit: 1, Embedder: constEmbedder(hotVector(0))})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 1 || results[0].Score == nil {
		t.Fatalf("expected scored result, got %+v", results)
	}
	if math.Abs(*results[0].Score-1.0) > 1e-6 {
		t.Fatalf("score = %f, want ~1.0 (distance 0)", *results[0].Score)
	}
}

func TestGetMemoryRecordLocation(t *testing.T) {
	d := newTestDB(t)
	id := insertRecord(t, d, memory(withLines(11, 22), withArchive("/archive/loc.jsonl"), withDedupe("loc")))

	loc, err := GetMemoryRecordLocation(d, id)
	if err != nil {
		t.Fatal(err)
	}
	if loc == nil || loc.ArchivePath != "/archive/loc.jsonl" || loc.LineStart != 11 || loc.LineEnd != 22 {
		t.Fatalf("loc = %+v", loc)
	}

	// Missing id -> nil.
	missing, err := GetMemoryRecordLocation(d, 99999)
	if err != nil {
		t.Fatal(err)
	}
	if missing != nil {
		t.Fatalf("expected nil for missing id, got %+v", missing)
	}

	// Superseded -> nil.
	supID := insertRecord(t, d, memory(withStatus(db.StatusSuperseded), withDedupe("sup-loc"), withArchive("/archive/sup.jsonl")))
	sup, err := GetMemoryRecordLocation(d, supID)
	if err != nil {
		t.Fatal(err)
	}
	if sup != nil {
		t.Fatalf("superseded location should be nil, got %+v", sup)
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

func endsWith(s, suf string) bool {
	return len(s) >= len(suf) && s[len(s)-len(suf):] == suf
}
